### PR TITLE
Enable initial r support

### PR DESCRIPTION
**Pull request by BlueDrink9**
I'm still wrapping my head around treesitter to add the supported scopes, so I'll update this PR and add relevant tests once I do + once the r PR is added to vscode's parse tree

## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [ ] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [x] I have not broken the cheatsheet

### CODE REVIEW

**Comment by BlueDrink9** [comment]
I'm marking this as ready to make adding scopes easier. Once r support is included in cursorless, it will be easier for it to pick it up as an option for adding scopes

**Comment by AndreasArvidsson** [comment]
Latest version of the parse tree is published and should be available in the marketplace very soon.

File: data/fixtures/scopes/r/functionName.scope [hunk 1]
```diff
@@ -0,0 +1,19 @@
+abc <- function(arg){
+    TRUE
+}
+---
+
+[Content] =
+[Domain] = 0:0-0:3
+  >---<
+0| abc <- function(arg){
```
Comment (BlueDrink9) [inline data/fixtures/scopes/r/functionName.scope:9]: This is the only case that I'm unsure about and is likely to need review, I'm wondering if it would make more sense to include the operator in this for removal and insertion? Although I might struggle a little bit with adding them as custom delimiters

**Comment by BlueDrink9** [comment]
@AndreasArvidsson ready to review! Thank you again for your help yesterday

**Comment by AndreasArvidsson** [comment]
Please add a scope test for an empty argument list as well. The difference is the insertion delimiter. Please have a look at how it's implemented in other languages.

File: data/fixtures/scopes/r/functionName.scope (hunk 1 above)
Comment (AndreasArvidsson) [reply data/fixtures/scopes/r/functionName.scope:9]: `<-` should probably be added as a trailing range. That way it will be deleted if you delete the name.
We do the same thing with `=` in `value = 2`

**Comment by AndreasArvidsson** [comment]
I would recommend that you create multiple scope files for the same facet instead of adding multiple different code examples in the same file. It's much simpler to review it that way.

File: data/fixtures/scopes/r/name.assignment.scope [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  >-----<
+0| hello <- "world"
+
+[#1 Removal] = 0:0-0:6
+  >------<
+0| hello <- "world"
+
+[#1 Trailing delimiter] = 0:5-0:6
+       >-<
+0| hello <- "world"
```
Comment (AndreasArvidsson) [inline data/fixtures/scopes/r/name.assignment.scope:16]: ` <- ` should be part off the trailing range. Have a look at how we do this in other languages.

File: data/fixtures/scopes/r/name.assignment.scope [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  >-----<
+1| hello = "world"
+
+[#2 Removal] = 1:0-1:6
+  >------<
+1| hello = "world"
+
+[#2 Trailing delimiter] = 1:5-1:6
+       >-<
+1| hello = "world"
```
Comment (AndreasArvidsson) [inline data/fixtures/scopes/r/name.assignment.scope:32]: ` = ` should be part of the trailing range

File: data/fixtures/scopes/r/name.variable.scope [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  >-----<
+0| hello <- "world"
+
+[#1 Removal] = 0:0-0:6
+  >------<
+0| hello <- "world"
+
+[#1 Trailing delimiter] = 0:5-0:6
+       >-<
+0| hello <- "world"
```
Comment (AndreasArvidsson) [inline data/fixtures/scopes/r/name.variable.scope:16]: ` <- ` should be part of the trailing range

File: data/fixtures/scopes/r/name.variable.scope [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  >-----<
+1| hello = "world"
+
+[#2 Removal] = 1:0-1:6
+  >------<
+1| hello = "world"
+
+[#2 Trailing delimiter] = 1:5-1:6
+       >-<
+1| hello = "world"
```
Comment (AndreasArvidsson) [inline data/fixtures/scopes/r/name.variable.scope:32]: ` = ` should be part of the trailing range

File: data/fixtures/scopes/r/value.assignment.scope [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+           >-------<
+0| hello <- "world"
+
+[#1 Removal] = 0:8-0:16
+          >--------<
+0| hello <- "world"
+
+[#1 Leading delimiter] = 0:8-0:9
+          >-<
+0| hello <- "world"
```
Comment (AndreasArvidsson) [inline data/fixtures/scopes/r/value.assignment.scope:16]: ` <- ` should be part of the leading range

File: data/fixtures/scopes/r/value.assignment.scope [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+          >-------<
+1| hello = "world"
+
+[#2 Removal] = 1:7-1:15
+         >--------<
+1| hello = "world"
+
+[#2 Leading delimiter] = 1:7-1:8
+         >-<
+1| hello = "world"
```
Comment (AndreasArvidsson) [inline data/fixtures/scopes/r/value.assignment.scope:32]: ` = ` should be part of the leading range

File: data/fixtures/scopes/r/value.variable.scope [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+           >-------<
+0| hello <- "world"
+
+[#1 Removal] = 0:8-0:16
+          >--------<
+0| hello <- "world"
+
+[#1 Leading delimiter] = 0:8-0:9
+          >-<
+0| hello <- "world"
```
Comment (AndreasArvidsson) [inline data/fixtures/scopes/r/value.variable.scope:16]: ` <- ` should be part of the leading range

File: data/fixtures/scopes/r/value.variable.scope [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+[#1 Insertion delimiter] = " "
+
+
+[#2 Content] =
+[#2 Domain] = 1:8-1:15
+          >-------<
+1| hello = "world"
```
Comment (AndreasArvidsson) [inline data/fixtures/scopes/r/value.variable.scope:24]: ` = ` should be part of the leading range

File: packages/common/src/scopeSupportFacets/r.ts [hunk 10]
```diff
@@ -0,0 +1,189 @@
+import type { LanguageScopeSupportFacetMap } from "./scopeSupportFacets.types";
+import { ScopeSupportFacetLevel } from "./scopeSupportFacets.types";
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const { supported, unsupported, notApplicable } = ScopeSupportFacetLevel;
+
+export const rScopeSupport: LanguageScopeSupportFacetMap = {
```
Comment (AndreasArvidsson) [inline packages/common/src/scopeSupportFacets/r.ts:7]: Could you please group these in supported/unsupported/notApplicable? Have a look at how we have done in other languages.

File: packages/common/src/scopeSupportFacets/r.ts [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+  "functionCallee": supported,
+
+  "name.assignment": supported,
+  "name.assignment.pattern": notApplicable,
+  "name.variable": supported,
+  "value.assignment": supported,
+  "value.variable": supported,
+  "value.variable.pattern": notApplicable,
+
+  // "interior.function": unsupported,
```
Comment (AndreasArvidsson) [inline packages/common/src/scopeSupportFacets/r.ts:28]: Before this is ready to merge I would like all commented facets like this to be removed

**Review by AndreasArvidsson — CHANGES_REQUESTED** [review]

File: data/fixtures/scopes/r/argument.actual.scope [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+       >-<
+0| func(x, y)
+
+[#1 Removal] = 0:5-0:8
+       >---<
+0| func(x, y)
+
+[#1 Trailing delimiter] = 0:6-0:8
+        >--<
+0| func(x, y)
```
[the remark this answers is not in the document]
Comment (BlueDrink9) [reply data/fixtures/scopes/r/argument.actual.scope:15]: Part of me wonders if the parser is the issue here, it seems like the capture after an argument does not include the white space, which it seems to me it it does for other languages?

[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply data/fixtures/scopes/r/argument.actual.scope:15]: The problem is that the parse tree have `,` as a named node. This is quite unusual. Usually delimiters like `,`, `(` etc are not named. To fix this we can do the following:

```scm
(
  (arguments
    (
      (argument) @_.leading.endOf
      (comma)
    )?
    .
    (argument) @argumentOrParameter
    .
    (
      (comma)
      (argument) @_.trailing.startOf
    )?
  ) @_dummy
  (#not-type? @argumentOrParameter "comment")
  (#single-or-multi-line-delimiter! @argumentOrParameter @_dummy ", " ",\n")
)
```

File: packages/cursorless-vscode/src/scripts/initLaunchSandbox.ts [hunk 13]
```diff
@@ -22,7 +22,7 @@ async function main() {
     }
 
     // Install extension dependencies
-    const subprocess = cp.spawn("code", args, {
+    const subprocess = cp.spawn("codium", args, {
```
Comment (AndreasArvidsson) [inline packages/cursorless-vscode/src/scripts/initLaunchSandbox.ts:25]: Please don't commit this

File: packages/cursorless-org-docs/src/docs/contributing/adding-a-new-language.md [hunk 14]
```diff
@@ -1,3 +1,25 @@
+See description of facets in here,
```
Comment (AndreasArvidsson) [inline packages/cursorless-org-docs/src/docs/contributing/adding-a-new-language.md:1]: This change needs to be worked on. It looks like a bunch of separate thoughts. Did you really mean to commit it like this?

File: packages/cursorless-vscode/src/scripts/initLaunchSandbox.ts (hunk 13 above)
Comment (BlueDrink9) [reply packages/cursorless-vscode/src/scripts/initLaunchSandbox.ts:25]: oops!

File: packages/cursorless-org-docs/src/docs/contributing/adding-a-new-language.md (hunk 14 above)
Comment (BlueDrink9) [reply packages/cursorless-org-docs/src/docs/contributing/adding-a-new-language.md:1]: No I did not, sorry! That was just my todo list, I am not used to how VSCode auto stages all changes

Comment (AndreasArvidsson) [reply packages/cursorless-org-docs/src/docs/contributing/adding-a-new-language.md:1]: Vscode doesn't auto stage anything for me. Might be a setting. That sounds dangerous :D

File: data/fixtures/scopes/r/argument.actual.scope (hunk 12 above)
[the remark this answers is not in the document]
Comment (BlueDrink9) [reply data/fixtures/scopes/r/argument.actual.scope:15]: Thanks, that worked!

[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply data/fixtures/scopes/r/argument.actual.scope:15]: Great. btw since we're not using the wild card anymore you can remove `  (#not-type? @argumentOrParameter "comment")`

File: data/fixtures/scopes/r/ifStatement2.scope [hunk 15]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (AndreasArvidsson) [inline data/fixtures/scopes/r/ifStatement2.scope:1]: Is there a particular reason for this leading empty line?

File: packages/common/src/scopeSupportFacets/r.ts [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+  "functionCallee": supported,
+  "functionName": supported,
+  "ifStatement": supported,
+  "name.assignment": supported,
+  "name.variable": supported,
+  "namedFunction": supported,
+  "value.assignment": supported,
+  "value.variable": supported,
+
+  "branch.if": unsupported,
```
Comment (AndreasArvidsson) [inline packages/common/src/scopeSupportFacets/r.ts:22]:
```suggestion
  // Unsupported

  "branch.if": unsupported,
```

File: packages/common/src/scopeSupportFacets/r.ts [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+  attribute: unsupported,
+  class: unsupported,
+  className: unsupported,
+  environment: unsupported,
+  fieldAccess: unsupported,
+  map: unsupported,
+  regularExpression: unsupported,
+  switchStatementSubject: unsupported,
+
+  "interior.cell": notApplicable,
```
Comment (AndreasArvidsson) [inline packages/common/src/scopeSupportFacets/r.ts:60]:
```suggestion
  // Not applicable

  "interior.cell": notApplicable,
```

File: packages/common/src/scopeSupportFacets/r.ts [hunk 18]
```diff
@@ -0,0 +1,96 @@
+import type { LanguageScopeSupportFacetMap } from "./scopeSupportFacets.types";
+import { ScopeSupportFacetLevel } from "./scopeSupportFacets.types";
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (AndreasArvidsson) [inline packages/common/src/scopeSupportFacets/r.ts:4]:
```suggestion
```

**Comment by AndreasArvidsson** [comment]
Looks like you're starting to get the hang of this :)

* Might be worth adding a test for a vertical argument list for both `argument.actual` and `argumentList.formal`.
* Looks like you're still missing `argument.formal`
* Don't forget to add iteration scopes for argument otherwise you can't do `every arg`

File: data/fixtures/scopes/r/argument.actual.iteration.scope [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+
+---
+
+[#1 Range] = 0:4-0:8
+      >----<
+0| foo(x, y)
+
+[#1 Domain] = 0:0-0:9
+  >---------<
+0| foo(x, y)
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply data/fixtures/scopes/r/argument.actual.iteration.scope:11]: The reason for this is when you use a function call as an argument. eg
```ts
foo(|bar(1), baz(2))
```
If your cursor is `|` and you say `"take every arg"` I would expect to get two selections with both bar and baz function calls. That is why we exclude the function name from the iteration domain.

File: data/fixtures/scopes/r/argument.formal.iteration.scope [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+
+---
+
+[#1 Range] = 0:16-0:20
+                  >----<
+0| abc <- function(x, y){}
+
+[#1 Domain] = 0:15-0:21
+                 >------<
+0| abc <- function(x, y){}
```
Comment (AndreasArvidsson) [inline data/fixtures/scopes/r/argument.formal.iteration.scope:11]: That domain should include the entire function

eg
https://github.com/cursorless-dev/cursorless/blob/97bf1d6651000087ea1762aec9bb5ca50bc1b00e/data/fixtures/scopes/javascript.core/argument.formal.iteration.scope

**Review by AndreasArvidsson — CHANGES_REQUESTED** [review]

File: data/fixtures/scopes/r/argument.formal.iteration.scope (hunk 20 above)
Comment (AndreasArvidsson) [reply data/fixtures/scopes/r/argument.formal.iteration.scope:11]: Have you read my response about that?
https://github.com/cursorless-dev/cursorless/pull/2721/files#r2106136228

Comment (AndreasArvidsson) [reply data/fixtures/scopes/r/argument.formal.iteration.scope:11]: You probably want to include the function name as well so that you can take all arguments from the function name

**Review by AndreasArvidsson — APPROVED** [review]

File: packages/cursorless-org-docs/src/docs/user/languages/r.mdx [hunk 21]
```diff
@@ -0,0 +1,5 @@
+import { Language } from "./components/Language";
+
+# r
```
Comment (AndreasArvidsson) [inline packages/cursorless-org-docs/src/docs/user/languages/r.mdx:3]:
```suggestion
# R
```

Comment (BlueDrink9) [reply packages/cursorless-org-docs/src/docs/user/languages/r.mdx]: The test expected r, how come?

Comment (BlueDrink9) [reply packages/cursorless-org-docs/src/docs/user/languages/r.mdx]: Oh right this is just markdown, nvm

Comment (AndreasArvidsson) [reply packages/cursorless-org-docs/src/docs/user/languages/r.mdx]: That was probably the file name. Them markdown title we are free to do whatever we want with.

**Comment by AndreasArvidsson** [comment]
@BlueDrink9 Regarding iteration domain for arguments
https://github.com/cursorless-dev/cursorless/issues/2930
